kubeflow/generate_train_eval.py

```diff
@@ -919,4 +922,1329 @@ def train_eval_sumbt(
+            dlg_side=dlg_side,
+            image=image,
+            genienlp_version=genienlp_version,
+            genie_version=genie_version,
+            thingtalk_version=thingtalk_version,
+            workdir_repo=workdir_repo,
+            workdir_version=workdir_version,
+            additional_args=additional_args
+        )
+        prepare_for_translation_op.container.set_image_pull_policy('Always')
+    
+    if do_translation:
+        do_translation_op = do_translation_step(
+            owner=owner,
+            project=project,
+            experiment=experiment,
+            s3_bucket=s3_bucket,
+            task_name=task_name,
+            s3_datadir=s3_datadir,
+            model_name_or_path=model_name_or_path,
+            input_splits=input_splits,
+            train_output_per_example=train_output_per_example,
+            nmt=nmt,
+            do_alignment=do_alignment,
+            src_lang=src_lang,
+            tgt_lang=tgt_lang,
+            dlg_side=dlg_side,
+            image=image,
+            genienlp_version=genienlp_version,
+            genie_version=genie_version,
+            thingtalk_version=thingtalk_version,
+            workdir_repo=workdir_repo,
+            workdir_version=workdir_version,
+            additional_args=additional_args
+        )
+        do_translation_op.container.set_image_pull_policy('Always')
+    
+        if prepare_for_translation:
+            do_translation_op.after(prepare_for_translation_op)
+    
+    if post_process_translation:
+        post_process_translation_op = post_process_translation_step(
+                owner=owner,
+                project=project,
+                experiment=experiment,
+                s3_bucket=s3_bucket,
+                task_name=task_name,
+                s3_datadir=s3_datadir,
+                model_name_or_path=model_name_or_path,
+                input_splits=input_splits,
+                train_output_per_example=train_output_per_example,
+                nmt=nmt,
+                do_alignment=do_alignment,
+                src_lang=src_lang,
+                tgt_lang=tgt_lang,
+                dlg_side=dlg_side,
+                image=image,
+                genienlp_version=genienlp_version,
+                genie_version=genie_version,
+                thingtalk_version=thingtalk_version,
+                workdir_repo=workdir_repo,
+                workdir_version=workdir_version,
+                additional_args=additional_args
+            )
+        post_process_translation_op.container.set_image_pull_policy('Always')
+    
+        if prepare_for_translation and do_translation:
+            post_process_translation_op.after(do_translation_op, prepare_for_translation_op)
+        elif do_translation:
+            post_process_translation_op.after(do_translation_op)
+
+@dsl.pipeline(
+    name='Prepare dataset for Translation',
+    description='Prepare English dataset for translation'
+)
+def prepare_for_translation(
```

Comment:
Why are these separate pipelines instead of one pipeline with 3 steps?

Reply:
The three steps don't always go together. `prepare_for_translation` only needs to run once per dataset. The results get translated and post-processed. I separated last two cause post-process sometimes fail and we can restart from translations (no need to re-do it). Also, only the translate step needs GPU so envs can be different.

Reply:
If you do cache right (which you should), you can rerun the whole pipeline without running all the steps that succeeded already.